File: modules/count.c
#include <stdlib.h>
#include "utone.h"

int ut_count_create(ut_count **p)
{
    *p = malloc(sizeof(ut_count));
    return UT_OK;
}

int ut_count_destroy(ut_count **p)
{
    free(*p);
    return UT_OK;
}

int ut_count_init(ut_data *ut, ut_count *p)
{
    p->count = 4;
    p->curcount = -1;
    p->mode = 0;
    return UT_OK;
}

int ut_count_compute(ut_data *ut, ut_count *p, UTFLOAT *in, UTFLOAT *out)
{
    if(*in){
        if(p->mode == 0) {
            p->curcount = (p->curcount + 1) % p->count;
        } else {
            if(p->curcount == -2) {
                *out = -2;
                return UT_OK;
            }
            if(p->curcount >= p->count - 1) {
                p->curcount = -2;
            } else {
                if(p->curcount == -1) p->curcount = 0;
                else p->curcount++;
            }
        }
    }
    *out = p->curcount;
    return UT_OK;
}

File: modules/data/count.lua
sptbl["count"] = {

    files = {
        module = "count.c",
        header = "count.h",
        example = "ex_count.c",
    },

    func = {
        create = "ut_count_create",
        destroy = "ut_count_destroy",
        init = "ut_count_init",
        compute = "ut_count_compute",
    },

    params = {
        optional = {
            {
                name = "count",
                type = "UTFLOAT",
                description = "Number to count up to (count - 1). Decimal points will be truncated.",
                default = 4
            },
            {
                name = "mode",
                type = "UTFLOAT",
                description = "Counting mode. 0 = wrap-around, 1 = count up to N -1, then stop and spit out -1",
                default = 0
            },
        },
    },

    modtype = "module",

    description = [[Trigger-based fixed counter

    The signal output will count from 0 to [N-1], and then
repeat itself. Count will start when it has been triggered, otherwise it will be -1.]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trig",
            description = "When non-zero, this value will increment."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Signal output."
        },
    }

}
